### PR TITLE
NVD Model

### PR DESCRIPTION
**Pull request by CermakM**
Create a new model for NVD Feed

Model is structured as a document with the idea of future document database usage.

Model has the following schema:

```python
    nvdlib.model.Document(
        cve: nvdlib.model.CVE(
             id_: str,
             assigner: str,
             data_version: str,
             affects: AffectsEntry(data: List[
                ProductNode(
                    vendor_name: str,
                    product_name: str,
                    versions: List[str])]
                )],
             references: ReferenceEntry(data: List[
                ReferenceNode(
                    url: str,
                    refsource: str
                )],
             descriptions: DescriptionEntry(data: List[
                DescriptionNode(
                    lang: str
                )]
        ),
        configurations: nvdlib.model.Configurations(
            cve_data_version: str,
            nodes: List[
                ConfigurationsEntry(
                    data: List[
                        ConfigurationsNode(
                            vulnerable: True,
                            cpe: str
                        )],
                    operator: str
            )]
        ),
        impact: nvdlib.model.Impact(
            severity: str,
            exploitability_score: float,
            impact_score: float,
            cvss: nvdlib.model.Impact.CVSSNode(
                version: str,
                access_vector: str,
                access_complexity: str,
                authentication: str,
                confidentiality_impact: str,
                integrity_impact: str,
                availability_impact: str,
                base_score: float
            )
        ),
        published_date: datetime.datetime(
            year: int,
            month: int,
            day: int,
            hour: int,
            minute: int
        ),
        modified_date: datetime.datetime(
            year: int,
            month: int,
            day: int,
            hour: int,
            minute: int
        )
    )
```

To test the precedure, I suggest the following approach:
``` =python
feed_names = [2017, 2018]

with FeedManager(n_workers=5) as fm:
    
    feeds = fm.fetch_feeds(feed_names)
    collection = fm.collect(feeds)

cursor = collection.cursor()
doc  = cursor.next()  # type: Document
batch = cursor.next_batch()  # type: List[Document]
```

### CODE REVIEW

**Comment by msrb** [comment]
@CermakM selector package is not part of the PR, but it is referenced from the code. Is that intentional, or just forgotten git-add? :)

**Comment by CermakM** [comment]
@msrb Keen eye! It is not supposed to be part of the PR, I totally forgot I used it for type prototyping. It shouldn't have impact on functionality, so either we can ignore this, or I can just replace the import statement with class proto. WDYT?

**Comment by msrb** [comment]
@CermakM depends... so what's the plan here? Because without the package, you cannot really test this PR as import will always fail.

**Comment by CermakM** [comment]
@msrb fixed

File: nvdlib/adapters/base.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    @abstractmethod
+    def cursor(self) -> BaseCursor:
+        """Initialize cursor to the beginning of a collection.
+
+        The cursor must have implemented at least the following functionality:
+            - next() ->
+            - batch_size()
+        """
+
+    def is_connected(self):
```
Comment (msrb) [inline nvdlib/adapters/base.py]: Does consumer of this API care about whether underlying storage is connected or not?

File: nvdlib/adapters/default.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+def release_lock(*fd):
+    """Release locks on given file descriptors and close them."""
+    for locked_fd in fd:
+        if locked_fd is None:  # this can happen with default initialization of an adapter
+            continue
+        fcntl.flock(locked_fd, fcntl.LOCK_UN)
+
+        # wait for lock to be released
+        time.sleep(0.1)
```
[the remark this answers is not in the document]
Comment (CermakM) [reply nvdlib/adapters/default.py]: Ye, that's magic :) .. I am just used to put this line whenever I release lock out of 'good practice' to 'make sure' lock has been released. Of course, 'make sure' is overrated term but at least the probability is much higher. :) Should do no harm, tho.

File: nvdlib/adapters/base.py (hunk 1 above)
Comment (CermakM) [reply nvdlib/adapters/base.py]: Agreed this method is probably redundant and is a remainder of some of my thought processes (I believe it is not even used in the code, will get rid of it)

About the storages, that's perhaps a longer discussion as it does make sense for our default adapter and for some potential ones, however, for some other, it doesn't. I believe we can leave it in place for now and let it undergo some testing, wait for what future brings regarding new adapters and we'll get to that then. WDYT?

Comment (msrb) [reply nvdlib/adapters/base.py]: Yup, sounds like a plan :wink:

**Comment by CermakM** [comment]
Thanks for review and merging!

**Pull request merged**